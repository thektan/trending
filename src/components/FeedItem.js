import React, {Component} from 'react';
import moment from 'moment';

/**
 * Feed item.
 *
 * @class FeedItem
 * @extends {Component}
 */
class FeedItem extends Component {
	/**
	 * Converts a timestamp into a formatted date.
	 *
	 * @param {string|number} time
	 * @return {string} Formatted date.
	 */
	getFormattedDateString(time) {
		return moment(time).fromNow();
	}

	/**
	 * Renders a list item of a feed.
	 *
	 * @return {string} Feed item markup.
	 */
	render() {
		const {commentCount, score, date, title, url} = this.props;

		const formattedDate = this.getFormattedDateString(date);

		return (
			<li className="feed-item-container">
				<a href={url}>
					<span className="feed-item-title">{title}</span>
				</a>

				<div className="feed-item-metadata-container">
					{score} points - {commentCount} comments - {formattedDate}
				</div>
			</li>
		);
	}
}

export default FeedItem;
