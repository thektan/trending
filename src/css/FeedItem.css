/* bright-green: #2FF4AF */

.feed-item {
	list-style: none;
	width: 33%;
}

.feed-item__container {
	color: #1A2B50;
	display: flex;
	padding-bottom: 20px;
}

.feed-item__link {
	word-break: break-all;
}

.feed-item__link, .feed-item__metadata, .feed-item__metadata a {
	color: #7C8795;
	font-size: 12px;
}

.feed-item__score {
	color: #1C5EFF;
	font-size: 16px;
	font-weight: 600;
	min-width: 50px;
	margin-right: 20px;
	text-align: right;
}

.feed-item__title {
	color: #1A2B50;
	font-size: 18px;
	font-weight: 600;
	line-height: 1.3;
}

.feed-item__title + .feed-item__link {
	margin-left: 5px;
}

.feed-item__title-link {
	display: block;
	line-height: 1.2;
	margin-bottom: 2px;
}

.feed-item__title-link:hover .feed-item__title {
	background-color: #2FF4AF;
}

.feed-item__title-link:visited .feed-item__title {
	color: #BCC3CC;
}
