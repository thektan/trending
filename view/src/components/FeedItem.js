import React, {Component} from 'react';
import Isvg from 'react-inlinesvg';
import iconDiscuss from '../images/icon-discuss.svg';
import iconPopular from '../images/icon-popular.svg';
import '../css/FeedItem.css';
import {
	getFormattedDateString,
	getURLDomainName,
	toCssClassName,
} from '../utils/util';

/**
 * Feed item.
 *
 * @class FeedItem
 * @extends {Component}
 */
class FeedItem extends Component {
	/**
	 * Renders a list item of a feed.
	 *
	 * @return {string} Feed item markup.
	 */
	render() {
		const {
			commentCount,
			discuss,
			score,
			date,
			siteUrl,
			sourceName,
			title,
			url,
			popular,
		} = this.props;

		const formattedDate = getFormattedDateString(date);

		return (
			<li className={'feed-item feed-item--' + toCssClassName(sourceName)}>
				<div className="feed-item__container">
					<div className="feed-item__score">
						{score}

						<div className="badge-container">
							{popular ?
								<div className="badge badge--popular">
									<Isvg className="icon" src={iconPopular}></Isvg>
								</div> : ''
							}

							{discuss ?
								<div className="badge badge--discuss">
									<Isvg className="icon" src={iconDiscuss}></Isvg>
								</div> : ''
							}
						</div>
					</div>

					<div className="feed-item__content">
						<a
							className="feed-item__title-link"
							href={url || siteUrl}
							target="_blank"
						>
							<span className="feed-item__title">
								{title}
							</span>

							<span className="feed-item__link">
								({getURLDomainName(url || siteUrl)})
							</span>
						</a>

						<div className="feed-item__metadata">
							<a href={siteUrl} target="_blank">
								{commentCount >= 0 ? commentCount + ' comments â€¢' : ''}
							</a> {formattedDate}
						</div>
					</div>
				</div>
			</li>
		);
	}
}

export default FeedItem;
