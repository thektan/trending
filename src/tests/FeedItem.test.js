import React from 'react';
import {shallow} from 'enzyme';
import renderer from 'react-test-renderer';

import MockDate from 'mockdate';

import FeedItem from '../components/FeedItem';

describe('FeedItem', () => {
	it('renders', () => {
		MockDate.set(1499108740959);

		const component = renderer.create(
			<FeedItem
				commentCount={1}
				date={1499094043 * 1000}
				score={1}
				siteUrl='https://testurl.com'
				title='Test Title'
				url='https://testurl.com'
			/>
		);

		const tree = component.toJSON();
		expect(tree).toMatchSnapshot();

		MockDate.reset();
	});
});

describe('getURLDomainName function', () => {
	const component = shallow(
		<FeedItem
			commentCount={1}
			date={1499094043 * 1000}
			score={1}
			siteUrl='https://testurl.com'
			title='Test Title'
			url='https://testurl.com'
		/>
	);

	const getURLDomainName = component.instance().getURLDomainName;

	it('handles basic url string', () => {
		expect(getURLDomainName('https://domain.com'))
			.toEqual('domain.com');
	});

	it('handles url with a subdomain', () => {
		expect(getURLDomainName('https://sub.domain.com'))
			.toEqual('sub.domain.com');
	});

	it('handles url with multiple parameters', () => {
		expect(getURLDomainName('https://domain.com/extra/pages?a=testing&b=another'))
			.toEqual('domain.com');
	});

	it('handles url with file type', () => {
		expect(getURLDomainName('http://domain.com/file.pdf'))
			.toEqual('domain.com');
	});

	it('handles url without http', () => {
		expect(getURLDomainName('domain.com'))
			.toEqual('domain.com');
	});
});
