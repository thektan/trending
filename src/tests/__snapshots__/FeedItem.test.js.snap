// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedItem renders 1`] = `
<li
  className="feed-item"
>
  <div
    className="feed-item__container"
  >
    <div
      className="comment-popup "
    >
      <span
        className="comment-popup__close-icon"
        onClick={[Function]}
      >
        x
      </span>
      <div
        className="comment-popup__body"
      />
    </div>
    <div
      className="feed-item__score"
    >
      1
    </div>
    <div
      className="feed-item__content"
    >
      <a
        className="feed-item__title-link"
        href="https://testurl.com"
      >
        <span
          className="feed-item__title"
        >
          Test Title
        </span>
        <span
          className="feed-item__link"
        >
          (
          testurl.com
          )
        </span>
      </a>
      <div
        className="feed-item__metadata"
      >
        <a
          href="https://testurl.com"
          onClick={[Function]}
        >
          1 comments â€¢
        </a>
         
        4 hours ago
      </div>
    </div>
  </div>
</li>
`;
